### ISSUE TITLE
ClusterClass patch implementation

### ISSUE THREAD

**Issue by sbueringer** [kubernetes-sigs/cluster-api#5460]
This issue is an umbrella issue for the ClusterClass patch implementation. It's a follow-up to the ClusterClass patch proposal amendment: #5142.

Note: this will evolve over time as the implementation progresses.

### Initial Prereqs / PRs

* [x] [✨ Add ClusterClass variable and patch types #5418](https://github.com/kubernetes-sigs/cluster-api/pull/5418)
* [x] [🌱 Move ClusterClass webhook to a separate package, add ClusterClass deletion validation #5266](https://github.com/kubernetes-sigs/cluster-api/pull/5266)
    
### Validating & Defaulting

* [x] ClusterClass: variable validation (.spec.variables.): https://github.com/kubernetes-sigs/cluster-api/issues/5612
* [x] Cluster: variable value validation and defaulting (.spec.topology.variables): https://github.com/kubernetes-sigs/cluster-api/issues/5612

* ClusterClass: patch validation (.spec.patches)
    * [x] Validations | ClusterClass | ClusterClassPatch and PatchDefinition: https://github.com/kubernetes-sigs/cluster-api/issues/5613

* [x] compatibility: (slightly separate topic, but mentioned here for completeness)
    * update: variables cannot be removed as long as they are used in Clusters (#5461)
    * update: schemas of variables used in Clusters must not be changed in an incompatible way (#5461)
        
### Applying patches

A prototype implementation can be found on this ([PR](https://github.com/kubernetes-sigs/cluster-api/pull/5458)). 

**Update**: The first PR which adds the patch engine and variable calculation is open now: https://github.com/kubernetes-sigs/cluster-api/pull/5534
**Update**: The second PR which adds the JSON implementation is open now: https://github.com/kubernetes-sigs/cluster-api/pull/5583
**Update**: both PRs have been merged now

Code walkthrough: 
* 21th October: Recording: https://VMware.zoom.us/rec/share/_rXbvEHmL9dDQ2a-6RDGw3Vg4h3xLtG2RGHzUQCFwNDaSjpdBt9SPMoJvE37ZFFo.r0fCWQwj1CjTqggo Passcode: Begf8me@
* 28th October: Recording: https://VMware.zoom.us/rec/share/Kmn9LIUQde-5oa15cZFxxyBIwOJGob0FBtT2vNVQUrFWdJQQj-3w7gozCKElh7A.4gUZOtXG-A_IIn7I Passcode: z*g$gdT8

Context: Currently the topology controller has the following high-level steps:
* `getBlueprint` (gets blueprint, i.e. ClusterClass and referenced templates)
* `getCurrentState` (gets current state of the Cluster and the resources of the managed topology)
* `computeDesiredState` (compute desired state based on blueprint and current state)
* `reconcileState` (reconciles desired state)

Some central ideas of the implementation:
* Apply patches at the end of `computeDesiredState` and before `reconcileState`
    * We want to apply the patches at the end of `computeDesiredState` as we want to provide some parts of the desired state (e.g. labels/annotations and version) as builtin variables as input for patches. 
* We don't want to tightly couple the implementation to "inline patches" so that it's relatively easy to extend the implementation with e.g. external patches in the future.

The implementation can be split up in the following parts:
* [x] A generic patch engine (https://github.com/kubernetes-sigs/cluster-api/pull/5534)
* [x] Calculation of builtin variables (https://github.com/kubernetes-sigs/cluster-api/pull/5534)
* [x] The inline patching implementation, which mainly consists of: (https://github.com/kubernetes-sigs/cluster-api/pull/5583)
    * evaluation of patch selectors
    * evaluation of patch values

### Testing:

We solved most of the testing via extensive unit tests on the PRs listed above.
We are also using variables/patches in our regular ClusterClass e2e tests, e.g.: https://github.com/kubernetes-sigs/cluster-api/pull/5689. e2e tests of providers with ClusterClass will also heavily leverage variables/patches to make them as flexible as the non-ClusterClass tests are, e.g. https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/2945.

/kind feature
/area topology

**Comment by fabriziopandini** [comment]
@sbueringer @killianmuldoon  can we close this issue now that patch are merged and the web hook work is completed?

**Comment by killianmuldoon** [comment]
#5752 is still outstanding here - let's close this issue once that one is merged.

**Comment by sbueringer** [comment]
I'll close the issue once Killians PR is merged and I'll create a follow-up low-prio issue for advanced template validation.

**Comment by sbueringer** [comment]
Killian's PR has been merged. I created a separate issue to track https://github.com/kubernetes-sigs/cluster-api/issues/5809.

So this issue can be closed.

Great job everyone!

/close
